// Copyright 2025 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

syntax = "proto3";

package grpc.task;

import "core/pkg/service/task/pb/task.proto";
import "google/protobuf/empty.proto";

option go_package = "github.com/synnaxlabs/synnax/pkg/api/grpc/task";

service TaskCreateService {
  rpc Exec(CreateRequest) returns (CreateResponse);
}

service TaskRetrieveService {
  rpc Exec(RetrieveRequest) returns (RetrieveResponse);
}

service TaskDeleteService {
  rpc Exec(DeleteRequest) returns (google.protobuf.Empty);
}

service TaskCopyService {
  rpc Exec(CopyRequest) returns (CopyResponse);
}

message CreateRequest {
  repeated service.task.Task tasks = 1;
}

message CreateResponse {
  repeated service.task.Task tasks = 1;
}

message RetrieveRequest {
  uint32 rack = 1;
  repeated uint64 keys = 2;
  repeated string names = 3;
  repeated string types = 4;
  bool include_status = 5;
}

message RetrieveResponse {
  repeated service.task.Task tasks = 1;
}

message DeleteRequest {
  repeated uint64 keys = 1;
}

message CopyRequest {
  uint64 key = 1;
  string name = 2;
  bool snapshot = 3;
}

message CopyResponse {
  service.task.Task task = 1;
}
